core/src/ids.rs

```diff
@@ -210,6 +210,34 @@ impl ProgramId {
         let argument = [SALT, code_id.as_ref(), salt].concat();
         hash(&argument).into()
     }
+
+    /// Generate ProgramId from given CodeId, MessageId and salt
+    pub fn generate_for_pallet(code_id: CodeId, salt: &[u8], message_id: MessageId) -> Self {
+        const SALT: &[u8] = b"program_for_pallet";
+
+        let argument = [message_id.as_ref(), SALT, code_id.as_ref(), salt].concat();
+        hash(&argument).into()
+    }
+
+    /// Generate ProgramId from given CodeId, MessageId, salt and nonce
+    pub fn generate_for_wasm(
+        code_id: CodeId,
+        salt: &[u8],
+        message_id: MessageId,
+        nonce: usize,
+    ) -> Self {
```

Comment:
```suggestion
    pub fn generate_from_program(
        code_id: CodeId,
        salt: &[u8],
        message_id: MessageId,
    ) -> Self {
```
And keep duplicate init error here if salt matches